settings: Add personal setting for marking resolved-topic notices as read.

saubhagya-patel opened the pull request:
<!-- Describe your pull request here.-->
This PR adds a new personal setting to help users configure whether they want the resolved-topic notices to be marked as read depending on their preference.

As the [issue states](https://github.com/zulip/zulip/issues/33599#:~:text=This%20logic%20should%20replace%20the%20existing%20behavior%20where%20users%20who%20sent%20a%20message%20in%20the%20topic%20have%20that%20notification%20marked%20as%20unread.):
> This logic should replace the existing behavior where users who sent a message in the topic have that notification marked as unread.

Initially, we marked the resolved-topic notices as read by checking if a user has participated (sent a message or reacted to a message) in the topic. This PR replaces this logic and aims to mark the resolved topic notifications as read based on the personal setting of the user.

CZO thread(s):
- [#feedback > unexpected unread behavior with resolved topics](https://chat.zulip.org/#narrow/channel/137-feedback/topic/unexpected.20unread.20behavior.20with.20resolved.20topics)
- [#api design > resolve topic notifications unread setting](https://chat.zulip.org/#narrow/channel/378-api-design/topic/resolve.20topic.20notifications.20unread.20setting)


Fixes: #33599.

---

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
  <summary>Screenshot of added setting under "Play animated images" in Settings / Preferences > Information.</summary>

  ![image](https://github.com/user-attachments/assets/8fe51058-b6b5-4cf1-91c3-343961a00495)
</details>

<details>
  <summary>Screenshots of resolved-topic notices marked as read/unread based on the setting.</summary>

  - Please note that to demonstrate read/unread behavior, I have selected the setting "Automatically mark messages as read" as "Never", so as to correctly display unread messages.
  - For followed/unfollowed topics, kindly see the visibility policy near the topic name in the message header.
  - When a user has set "Immediately mark as read automated notices about topics being resolved" -> "Always":
    - Unfollowed topic  
      ![image](https://github.com/user-attachments/assets/375e89de-9734-4f45-8f06-bb95f8149893)

    - Followed topic  
      ![image](https://github.com/user-attachments/assets/374d10ae-e177-477a-8436-dcfa3c9c20b1)

  - When a user has set "Immediately mark as read automated notices about topics being resolved" -> "Except in topics I'm following":
    - Unfollowed topic  
      ![image](https://github.com/user-attachments/assets/d4455f6a-321f-4f51-96d9-e3f93e3f3ea3)

    - Followed topic  
      ![image](https://github.com/user-attachments/assets/6fcb2712-3c19-456a-b5f4-71c4fc13641b)

  - When a user has set "Immediately mark as read automated notices about topics being resolved" -> "Never":
    - Unfollowed topic  
      ![image](https://github.com/user-attachments/assets/9ad2a49d-b3fe-46d8-a469-73b1f4aa1d9e)

    - Followed topic  
      ![image](https://github.com/user-attachments/assets/2ea56f1c-23d1-4d61-9388-543332e5c72c)
</details>





---


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

saubhagya-patel commented on zerver/actions/message_edit.py:
```diff
@@ -162,11 +161,50 @@ def validate_user_can_edit_message(
+    stream: Stream,
+    topic_name: str,
+) -> set[int]:
+    all_stream_user_profiles = UserProfile.objects.filter(
+        id__in=Subscription.objects.filter(
+            recipient__type=Recipient.STREAM,
+            recipient__type_id=stream.id,
+            active=True,
+        ).values_list("user_profile_id", flat=True)
```
I am not entirely sure if this is the correct approach, but `active=True` is used here as I believe it will improve performance.

timabbott replied:
Use `get_active_subscriptions_for_stream_id`, which is a helper for this. I think you'll want to exclude deactivated users here.

saubhagya-patel replied:
Sure, I have updated `get_unread_user_ids_for_resolved_topic_notifications()` [function](https://github.com/zulip/zulip/pull/34226/files#diff-26aa930b41caa2a7e31ee5622de9b7629fbe56f84c27a87315fa8189420d45e2R164) to use the `get_active_subscriptions_for_stream_id()` helper.

saubhagya-patel commented on web/templates/settings/preferences_information.hbs:
```diff
@@ -49,13 +49,20 @@
         </div>
     </div>
 
-    <div class="input-group thinner setting-next-is-related">
```
These classes should be applied to the "Show unread counts for" setting, as the next setting, " Show counts for starred messages" is related to it.

Should I undo the changes or apply them to the "Show unread counts for" setting?

timabbott replied:
Good find. Probably move them to the intended thing, in a separate PR with before/after screenshots for review, since it's independent from these other changes?

saubhagya-patel replied:
Opened a PR #34389 to correctly apply the `thinner` class.

alya commented:
@prakhar1144 Could you please review this one? I haven't tested.

IEC61960 commented:
Form.support-content-visit-id
01743868134920-8811276398010228031

IEC61960 commented:
Form.support-content-visit-id
01743868134920-8811276398010228031
[Uploading lnc.com-statement (1).HTML…]()

timabbott commented on api_docs/unmerged.d/ZF-e0d354:
```diff
@@ -0,0 +1,5 @@
+  /settings`](/api/update-settings), [`PATCH
+  /realm/user_settings_defaults`](/api/update-realm-user-settings-defaults):
+  Added new `mark_resolved_topic_notifications_as_read_policy` setting, which
+  controls how resolved topic notifications are marked as read for a user.
```
This is missing a newline at the end of the file. 

@Vector73 I think this reflects a bug in our tooling -- we should give these a `.md` extension, so the Markdown linter runs on them.

Vector73 replied:
Opened #34386 to fix this.

saubhagya-patel replied:
Added the missing newline at the end of the file.
#34386 is now merged. Hence, `ZF-e0d354` is converted to [`ZF-e0d354.md`](https://github.com/zulip/zulip/pull/34226/files#diff-0b1ec6a18e14ccd4200637d7ac21bf7fb33437c2826d6f748398bff5f563053a).

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -162,11 +161,50 @@ def validate_user_can_edit_message(
+            user_profile_id__in=all_stream_user_ids,
+            visibility_policy=UserTopic.VisibilityPolicy.FOLLOWED,
+        ).values_list("user_profile_id", flat=True)
+    )
```
Ideally we'd not do a separate database query to fetch this -- don't we already fetch this exact data set elsewhere in the message-send code path?

(It might be too messy to share; if so, we should add a comment at least)

saubhagya-patel replied:
Yes, we do fetch this data.
I have updated the `get_unread_user_ids_for_resolved_topic_notifications()` [function](https://github.com/zulip/zulip/pull/34226/files#diff-26aa930b41caa2a7e31ee5622de9b7629fbe56f84c27a87315fa8189420d45e2R164) to share the data.
A [comment](https://github.com/zulip/zulip/pull/34226/files#diff-26aa930b41caa2a7e31ee5622de9b7629fbe56f84c27a87315fa8189420d45e2R1134-R1136) is also added.

timabbott commented on zerver/actions/realm_settings.py:
```diff
@@ -518,6 +523,8 @@ def do_set_realm_user_default_setting(
         value=value,
     )
+    if isinstance(raw_value, Enum):
+        event["value"] = raw_value.name
```
It'd be better to instead define an `event_value` field above in your existing if/else statement.

timabbott commented on zerver/actions/user_settings.py:
```diff
@@ -480,6 +489,9 @@ def do_change_user_setting(
         event["language_name"] = get_language_name(setting_value)
 
+    if isinstance(raw_setting_value, Enum):
+        event["value"] = raw_setting_value.name
```
Similarly, do an `event_value` here.

timabbott commented on zerver/tests/test_message_move_topic.py:
```diff
@@ -1737,20 +1723,176 @@ def test_mark_topic_as_resolved(self) -> None:
         assert (
             UserMessage.objects.filter(
-                user_profile__in=[admin_user, hamlet, aaron], message__id=messages[4].id
+                user_profile__in=[cordelia, hamlet], message__id=messages[1].id
```
This test feels hard to read. I suggest a prep commit adding a helper function for this check, and having it return the full set of user IDs who have an unread UserMessage for that message ID.

Maybe even in `test_classes.py` -- `get_user_ids_for_whom_message_unread(message_id)`.

And then the ~10 lines of code you have here for checking that can just be a call to that function and asserting the set of IDs it returns... and you can use the space in the test to make the comments super clear about why each block has the result it shows.

saubhagya-patel replied:
Ooh, apologies for the inconvenience caused.
The test [`test_resolved_topic_notice_unread_policy()`](https://github.com/zulip/zulip/pull/34226/files#diff-e9f5f5204b3e5f4ed4283822636ff59a4579c2cc67b614ad7676d0ba70a6fe5dR1726) is now updated, with an explanation of the result through comments in the following push.

> I suggest a prep commit adding a helper function for this check, and having it return the full set of user IDs who have an unread UserMessage for that message ID.
Maybe even in `test_classes.py` -- `get_user_ids_for_whom_message_unread(message_id)`.

Added a prep commit: [test_classes: Add functions to get read/unread user IDs for a message.](https://github.com/zulip/zulip/pull/34226/commits/b615bf8cfe8a4205f144e10d6688fcb8ddf487a6)

timabbott commented on zerver/lib/test_classes.py:
```diff
@@ -1197,6 +1197,22 @@ def get_messages(
+            .extra(where=[UserMessage.where_unread()])  # noqa: S610
+            .values_list("user_profile_id", flat=True)
+        )
+        return user_ids
```
Usually I'd write this commit in a way that it also migrates the code to use it. If you do that, I can merge the two prep commits to shrink this PR -- but I can't as you've done it, since this commit will fail CI.

saubhagya-patel replied:
Sure, I have updated the prep commit: [`test_classes: Add functions to get read/unread user IDs for a message.`](https://github.com/zulip/zulip/pull/34226/commits/c4e8f9de3f1164240fc0501e4b6e9429a4640f97).

Definitely a good lesson, will remember this in the future.
Thank you.

saubhagya-patel commented on zerver/openapi/zulip.yaml:
```diff
@@ -12861,6 +12861,21 @@ paths:
+                  description: |
+                    Controls whether the resolved topic notification is marked as unread.
+
+                    - "always" - Always mark resolved topic notifications as unread.
```
Sure, I have updated this in the following push. Thank you.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/34226#discussion_r2046020304) comment, updated the PR to use the terminology `resolved-topic notices` instead of `resolved topic notifications`.

- Following [this](https://github.com/zulip/zulip/pull/34226#discussion_r2047475489) comment, refactored the commit: [test_classes: Add functions to get read/unread user IDs for a message.](https://github.com/zulip/zulip/pull/34226/commits/c4e8f9de3f1164240fc0501e4b6e9429a4640f97), so that it also migrates the code using the functions. 

- <details>
  <summary> Updated label in settings. </summary>

  ![image](https://github.com/user-attachments/assets/a8199988-18d8-4fce-a9f3-d0cf177ec167)

  </details>

Thank you.

alya commented:
I don't think "resolved-topic notices" in good language for the settings context. (I have no comment on the API docs.)

I much prefer the original framing; I don't think keeping them unread is the right mental model.

alya commented:
"Immediately mark as read automated notices about topics being resolved" could work, if we want to spell it out.

timabbott commented:
Looks like #34389 auto-closed this because of GitHub's buggy regexes. Reopening: `Fixes: [#34226 (comment)](https://github.com/zulip/zulip/pull/34226#discussion_r2026449812).`

timabbott commented:
The challenge with the original framing is "In topics I'm not following" feels awkward. I don't think "Keep" is right though; we could do "Mark notices as unread", or something... but maybe there's a better answer for how to describe the "non-followed" case. Like "Except in followed topics."?

alya commented:
Using the strings specified in the issue but replacing  "In topics I'm not following" -> "Except in followed topics" sounds good to me.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes were made:

- Following the [CZO](https://chat.zulip.org/) thread: [#api design > resolve topic notifications unread setting](https://chat.zulip.org/#narrow/channel/378-api-design/topic/resolve.20topic.20notifications.20unread.20setting), the setting `resolved_topic_notice_unread_policy` with values [`always`, `unfollowed_topics`, `never`] is updated to `resolved_topic_notice_auto_read_policy` with values [`always`, `except_followed`, `never`].

- Following the comments ([[1]](https://github.com/zulip/zulip/pull/34226#issuecomment-2816215245), [[2]](https://github.com/zulip/zulip/pull/34226#issuecomment-2816219564), [[3]](https://github.com/zulip/zulip/pull/34226#issuecomment-2821961650), [[4]](https://github.com/zulip/zulip/pull/34226#issuecomment-2823348202), and [CZO thread](https://chat.zulip.org/#narrow/channel/378-api-design/topic/resolve.20topic.20notifications.20unread.20setting)) the setting labels have been updated:
  - Setting: "Immediately mark as read automated notices about topics being resolved".
  - Options: Always, Except in topics I'm following[default], Never.
  - <details>
    <summary>Screenshot of added setting with updated labels.</summary>

    ![image](https://github.com/user-attachments/assets/8fe51058-b6b5-4cf1-91c3-343961a00495)
    </details>

Thank you.

timabbott commented on zerver/actions/user_settings.py:
```diff
@@ -433,19 +434,32 @@ def update_scheduled_email_notifications_time(
     user_profile: UserProfile,
     setting_name: str,
-    setting_value: bool | str | int,
+    raw_setting_value: bool | str | int | Enum,
```
Hmm, do you think it's really worth renaming this to `raw_setting_value`, given how much it contaminates callers? 

I feel like we could just use `db_setting_value/setting_value` instead of `setting_value/raw_setting_value` and not need to churn and make callers less readable.

saubhagya-patel replied:
Yeah, I agree it's not worth renaming `setting_value` to `raw_setting_value`.
I have updated the function in the following push to use `db_setting_value/setting_value` instead of `setting_value/raw_setting_value`.
Thank you.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -1113,6 +1144,12 @@ def exclude_duplicates_from_subscription(
+    # `resolved_topic_notice_auto_read_policy` user setting.
+    user_profile_to_topic_visibility_policy: dict[UserProfile, int] = {}
+    if moved_all_visible_messages:
+        user_profile_to_topic_visibility_policy = orig_topic_user_profile_to_visibility_policy
```
(And that conditional fetch can be more tactical -- we only need to fetch the set of `user_ids` where the `UserTopic` is followed and they're in the channel. I think it might be cleaner to just pass a set of `users_following_original_topic` rather than the whole `orig_topic_user_profile_to_visibility_policy` as well.

saubhagya-patel replied:
Sure, I have updated the functions following the above review.

However, I have a suggestion, can we pass `user_ids_following_original_topic` instead of `users_following_original_topic`, then the function will become quite clean. 
Here's the diff (I have only changed the `get_unread_user_ids_for_resolved_topic_notifications()` function for keeping the diff concise):

```diff
diff --git a/zerver/actions/message_edit.py b/zerver/actions/message_edit.py
index 15740c9c03..489179b6a2 100644
--- a/zerver/actions/message_edit.py
+++ b/zerver/actions/message_edit.py
@@ -177,7 +177,7 @@ def validate_user_can_edit_message(
 def get_unread_user_ids_for_resolved_topic_notifications(
     stream: Stream,
     topic_name: str,
-    users_following_topic: set[UserProfile] | None,
+    user_ids_following_topic: set[int] | None,
 ) -> set[int]:
     unread_user_ids: set[int] = set()
 
@@ -188,10 +188,7 @@ def get_unread_user_ids_for_resolved_topic_notifications(
         is_bot=False,
     )
 
-    user_ids_following_topic: set[int] = set()
-    if users_following_topic is not None:
-        user_ids_following_topic = {user.id for user in users_following_topic}
-    else:
+    if user_ids_following_topic is None:
         user_ids_following_topic = set(
             UserTopic.objects.filter(
                 stream=stream,
```

timabbott commented on web/templates/settings/preferences_information.hbs:
```diff
@@ -56,6 +56,13 @@
     </div>
 
+    <div class="input-group">
+        <label for="{{prefix}}resolved_topic_notice_auto_read_policy" class="settings-field-label">{{t "Immediately mark as read automated notices about topics being resolved" }}</label>
```
I don't think this string is what the issue called for?

saubhagya-patel replied:
Yes, the issue suggests the string: "Automatically mark resolved topic notifications as read".
This string was suggested by @alya in [this](https://github.com/zulip/zulip/pull/34226#issuecomment-2816219564) comment.
Should we change it to follow the string from issue #33599?

timabbott replied:
I think that was a less preferred alternative option. I ended up moving it and reverting to the original string from the issue, except using the "notices" terminology.

``` diff
diff --git a/web/src/settings_notifications.ts b/web/src/settings_notifications.ts
index 9215d10a41..216ae19401 100644
--- a/web/src/settings_notifications.ts
+++ b/web/src/settings_notifications.ts
@@ -252,6 +252,10 @@ export function set_up(settings_panel: SettingsPanel): void {
         settings_object.automatically_follow_topics_policy,
     );
 
+    $container
+        .find(".setting_resolved_topic_notice_auto_read_policy")
+        .val(settings_object.resolved_topic_notice_auto_read_policy);
+
     const $automatically_unmute_topics_in_muted_streams_policy_dropdown = $container.find(
         ".setting_automatically_unmute_topics_in_muted_streams_policy",
     );
diff --git a/web/src/settings_preferences.ts b/web/src/settings_preferences.ts
index c695185ac7..4a5bce94fb 100644
--- a/web/src/settings_preferences.ts
+++ b/web/src/settings_preferences.ts
@@ -245,9 +245,6 @@ export function set_up(settings_panel: SettingsPanel): void {
     $container
         .find(".setting_web_animate_image_previews")
         .val(settings_object.web_animate_image_previews);
-    $container
-        .find(".setting_resolved_topic_notice_auto_read_policy")
-        .val(settings_object.resolved_topic_notice_auto_read_policy);
     $container
         .find(".setting_web_stream_unreads_count_display_policy")
         .val(settings_object.web_stream_unreads_count_display_policy);
diff --git a/web/templates/settings/notification_settings.hbs b/web/templates/settings/notification_settings.hbs
index 07f018ae72..2a713f23ea 100644
--- a/web/templates/settings/notification_settings.hbs
+++ b/web/templates/settings/notification_settings.hbs
@@ -96,6 +96,13 @@
                 {{> dropdown_options_widget option_values=automatically_unmute_topics_in_muted_streams_policy_values}}
             </select>
         </div>
+
+        <div class="input-group">
+            <label for="{{prefix}}resolved_topic_notice_auto_read_policy" class="settings-field-label">{{t "Automatically mark resolved topic notices as read" }}</label>
+            <select name="resolved_topic_notice_auto_read_policy" class="setting_resolved_topic_notice_auto_read_policy prop-element settings_select bootstrap-focus-style" id="{{prefix}}resolved_topic_notice_auto_read_policy" data-setting-widget-type="string">
+                {{> dropdown_options_widget option_values=resolved_topic_notice_auto_read_policy_values}}
+            </select>
+        </div>
     </div>
 
     <div class="desktop_notifications m-10 {{#if for_realm_settings}}settings-subsection-parent{{else}}subsection-parent{{/if}}">
diff --git a/web/templates/settings/preferences_information.hbs b/web/templates/settings/preferences_information.hbs
index dd58826764..4af8f114bb 100644
--- a/web/templates/settings/preferences_information.hbs
+++ b/web/templates/settings/preferences_information.hbs
@@ -56,13 +56,6 @@
         </select>
     </div>
 
-    <div class="input-group">
-        <label for="{{prefix}}resolved_topic_notice_auto_read_policy" class="settings-field-label">{{t "Immediately mark as read automated notices about topics being resolved" }}</label>
-        <select name="resolved_topic_notice_auto_read_policy" class="setting_resolved_topic_notice_auto_read_policy prop-element settings_select bootstrap-focus-style" id="{{prefix}}resolved_topic_notice_auto_read_policy" data-setting-widget-type="string">
-            {{> dropdown_options_widget option_values=resolved_topic_notice_auto_read_policy_values}}
-        </select>
-    </div>
-
     <div class="input-group">
         <label for="{{prefix}}web_stream_unreads_count_display_policy" class="settings-field-label">{{t "Show unread counts for" }}</label>
         <select name="web_stream_unreads_count_display_policy" class="setting_web_stream_unreads_count_display_policy prop-element bootstrap-focus-style settings_select" id="{{prefix}}web_stream_unreads_count_display_policy"  data-setting-widget-type="number">
```

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/34226#discussion_r2070940414) comment, the function `do_change_user_setting()` is updated in the following push to use `db_setting_value/setting_value` instead of `setting_value/raw_setting_value`.

- Following the comments [[1]](https://github.com/zulip/zulip/pull/34226#discussion_r2070943811), [[2]](https://github.com/zulip/zulip/pull/34226#discussion_r2070944768):
  - We now pass `users_following_original_topic` instead of `orig_topic_user_profile_to_visibility_policy` for getting unread ids for resolved topic notifications.
  - Added a conditional fetch when `users_following_original_topic` is `None` in the `maybe_send_resolve_topic_notifications()` function.

Please let me know if any further adjustments are required.
Thank you.

timabbott commented on zerver/migrations/0703_realmuserdefault_resolved_topic_notice_auto_read_policy_and_more.py:
```diff
@@ -0,0 +1,23 @@
+            name="resolved_topic_notice_auto_read_policy",
+            field=models.PositiveSmallIntegerField(default=except_followed),
+        ),
+    ]
```
This can be slow in very large installations aka Cloud; we can use `db_default` to avoid that.

``` diff
diff --git a/zerver/migrations/0703_realmuserdefault_resolved_topic_notice_auto_read_policy_and_more.py b/zerver/migrations/0703_realmuserdefault_resolved_topic_notice_auto_read_policy_and_more.py
index c29f5b65e0..a7874d5d9a 100644
--- a/zerver/migrations/0703_realmuserdefault_resolved_topic_notice_auto_read_policy_and_more.py
+++ b/zerver/migrations/0703_realmuserdefault_resolved_topic_notice_auto_read_policy_and_more.py
@@ -13,11 +13,17 @@ class Migration(migrations.Migration):
         migrations.AddField(
             model_name="realmuserdefault",
             name="resolved_topic_notice_auto_read_policy",
-            field=models.PositiveSmallIntegerField(default=except_followed),
+            field=models.PositiveSmallIntegerField(
+                default=except_followed,
+                db_default=except_followed,
+            ),
         ),
         migrations.AddField(
             model_name="userprofile",
             name="resolved_topic_notice_auto_read_policy",
-            field=models.PositiveSmallIntegerField(default=except_followed),
+            field=models.PositiveSmallIntegerField(
+                default=except_followed,
+                db_default=except_followed,
+            ),
         ),
     ]
diff --git a/zerver/models/users.py b/zerver/models/users.py
index a733f01ad4..a8e82dd0e3 100644
--- a/zerver/models/users.py
+++ b/zerver/models/users.py
@@ -272,7 +272,8 @@ class UserBaseSettings(models.Model):
     automatically_follow_topics_where_mentioned = models.BooleanField(default=True)
 
     resolved_topic_notice_auto_read_policy = models.PositiveSmallIntegerField(
-        default=ResolvedTopicNoticeAutoReadPolicyEnum.except_followed.value
+        default=ResolvedTopicNoticeAutoReadPolicyEnum.except_followed.value,
+        db_default=ResolvedTopicNoticeAutoReadPolicyEnum.except_followed.value,
     )
     RESOLVED_TOPIC_NOTICE_AUTO_READ_POLICY_TYPES = list(ResolvedTopicNoticeAutoReadPolicyEnum)
 
```

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -12999,6 +12999,21 @@ paths:
+                    - "except_followed" - Mark resolved-topic notices as read in topics not followed by the user.
+                    - "never" - Never mark resolved-topic notices as read.
+
+                    **Changes**: New in Zulip 11.0 (feature level ZF-e0d354.md).
```
You need it to say `New in Zulip 11.0 (feature level ZF-e0d354)`, without the `md`, for `tools/merge-api-changelogs` to work :). I fixed that before running it.

timabbott commented:
This is great, merged after the small edits noted above, thanks @saubhagya-patel! 

We need /help/ article updates for this; @alya probably should do this one, since it's a but involved.

alya commented:
Help center PR: #34632
